File: eftl/samples/python/publisher.py
'''
 Copyright (c) 2013-2020 TIBCO Software Inc.
 All Rights Reserved.
 For more information, please contact:
  TIBCO Software Inc., Palo Alto, California, USA
'''

import asyncio
import getopt, sys, time
import datetime
from messaging.eftl.connection import Eftl

user       = "user"
password   = "password"
url        = "ws://localhost:8585/channel"
connection = None
loop       = None
sendCount  = 10
rate       = 1
eFTL       = Eftl()
error      = False
sent       = 0

def usage():
    print (usage_lines)
    sys.exit(0)

usage_lines = """
        Usage: python3 publisher.py [options] url
        Options:
        \t-user        <str>    Username for authentication
        \t-password    <str>    Password for authentication
        \t-count       <int>    Number of message to publish
        \t-rate        <int>    Number of message per second
      """

def parseArgs(argv):
    count = len(argv)
    i = 1

    while i < count:
        if (argv[i].lower() == '-user'):
            global user
            user = argv[i+1]
        elif (argv[i].lower() == "-password"):
            global password
            password = argv[i+1]
        elif (argv[i].lower() == "-count"):
            global sendCount
            sendCount = int(argv[i+1])
        elif (argv[i].lower() == "-rate"):
            global rate
            rate = int(argv[i+1])
        elif (argv[i].lower() == "-help") or (argv[i].lower() == "-h"):
            usage()
        else:
           global url
           url = argv[i]

        i += 1

    print("publisher.py: " + eFTL.get_version())


# internal disconnect function
async def _disconnect():
    global connection

    if connection.is_connected():
        await connection.disconnect()

async def on_publish_complete(message):
    global sent
    print("Successfully Sent message [%s]: {%s}" % (sent, message))
    sent += 1

    if sent == sendCount:
        done = True
        print("Total messages sent: %s" % sent)

        if connection.is_connected():
            await _disconnect()

async def on_publish_error(message, code, reason):
    print("Failed to send message: %s, code: %s, reason: %s" %(message, code, reason))
    
    if connection.is_connected():
        await _disconnect()


# connection callback on_connect called on a successful connect
async def on_connect(connection):
    print("Connected to eFTL server: %s" % url)

# connection callback on_reconnect 
async def on_reconnect(connection):
    print("Reconnect attempt: ")

# connection callback on_error called if an error on the connection 
async def on_error(connection, code, reason):
    global error
    print("Connection error [%s], reason %s" %(code, reason))
    error = True

# connection callback on_disconnect called upon a disconnect from eFTL
async def on_disconnect(connection, loop, code, reason):
    print("Disconnected [%s]: %s" %(code, reason))

    loop.stop()

async def run():
    global connection
    global sendCount

    try:
        # connect to the eftl 
        connection = await eFTL.connect(url,
                                    user=user,
                                    password=password,
                                    trust_all=True,
                                    event_loop=loop,
                                    on_connect=on_connect,
                                    on_reconnect=on_reconnect,
                                    on_disconnect=on_disconnect,
                                    on_error=on_error)

        for i in range(sendCount):
            msg = connection.create_message()
            msg.set_string("type", "hello")
            msg.set_string("text", "This is a sample eFTL message")
            msg.set_long("long", i)
            msg.set_datetime("time", datetime.datetime.now())

            await connection.publish(msg,
                                     on_complete=on_publish_complete,
                                     on_error=on_publish_error)
           
            if rate > 0:
                await asyncio.sleep(float(1/rate))
            else:
                await asyncio.sleep(1)

    finally:
        print("Done..")

def cleanup():
    global loop
    global connection

    if connection.is_connected():
        loop.run_until_complete(_disconnect())

    loop.stop()
    loop.close()
    
def main(argv):
    errors = None
    global loop

    parseArgs(argv)

    try:
        loop = asyncio.get_event_loop();
        loop.run_until_complete(run())
        loop.run_forever()
    except (EOFError, KeyboardInterrupt):
        print("Disconnected from eFTL server: %s" % url)
    except CancelledError:
        logging.info('CancelledError')
    except Exception as err:
        errors = True
    finally:
           cleanup()
           if errors:
               sys.exit(1)
           else:
               sys.exit(0)

if __name__ == "__main__":
    main(sys.argv)

File: eftl/samples/dotnet/Publisher.cs
/*
 * Copyright (c) 2013-$Date: 2015-07-31 16:05:02 -0700 (Fri, 31 Jul 2015) $ Cloud Software Group, Inc.
 * All Rights Reserved.
 */

using System;
using System.Threading;
using System.Collections;

using TIBCO.EFTL;

public class Publisher
{
    static IConnection connection = null;

    static String url = "ws://localhost:8585/channel";
    static String username = "user";
    static String password = "password";
    static int count = 10;
    static int rate = 1;

    static CountdownEvent latch = new CountdownEvent(1);
    static Hashtable options = new Hashtable();
        
    private static void Usage() 
    {
        Console.WriteLine();
        Console.WriteLine("usage: Publisher [options] url");
        Console.WriteLine();
        Console.WriteLine("options:");
        Console.WriteLine("  -u, --username <username>");
        Console.WriteLine("  -p, --password <password>");
        Console.WriteLine("  -c, --count <count>");
        Console.WriteLine("  -r, --rate <messages per second>");
        Console.WriteLine();
        System.Environment.Exit(1);
    }

    public class ConnectionListener: IConnectionListener
    {
        public void OnConnect(IConnection connection)
        {
            Console.WriteLine("Connected to eFTL server at " + url);

            Publisher.connection = connection;
            Publisher.latch.Signal();
        }

        public void OnDisconnect(IConnection connection, int code, String reason)
        {
            Console.WriteLine("Disconnected from eFTL server: " + reason);
            Publisher.latch.Signal();
        }

        public void OnError(IConnection connection, int code, String reason)
        {
            Console.WriteLine("Connection error: " + reason);
            Publisher.latch.Signal();
        }

        public void OnReconnect(IConnection connection)
        {
            Console.WriteLine("Reconnected to eFTL server at " + url);
        }
    } 

    public class CompletionListener: ICompletionListener
    {
         public void OnCompletion(IMessage message) {
             Console.WriteLine("Published message " + message);
         }
    
         public void OnError(IMessage message, int code, String reason) {
             Console.WriteLine("Error publishing message: {0}\n", reason);
         }
    } 


    public Publisher(String url, String username, String password)
    {
        if (username != null)
            options.Add(EFTL.PROPERTY_USERNAME, username);

        if (password != null)
            options.Add(EFTL.PROPERTY_PASSWORD, password);
        
        // Asynchronously connect to the eFTL server.
        EFTL.Connect(url, options, new ConnectionListener());

        try {
            latch.Wait();
        }
        catch (Exception e) 
        {
            Console.WriteLine(e.Message);
        }
    }

    public bool IsConnected()
    {
        return connection.IsConnected();
    }

    public void Publish(int count, int rate)
    {
        for (int i = 1; i <= count; i++)
        {
            IMessage msg = connection.CreateMessage();

            // Publish "hello" messages. 
            //
            // When connected to an EMS channel the destination field must
            // be present and set to the EMS topic on which to publish the 
            // message.
            //
            msg.SetString("type", "hello");
            msg.SetString("text", "This is a sample eFTL message");
            msg.SetLong("long", (long) i);
            msg.SetDateTime("time", new DateTime());
            
            // Asynchronously publish the message.
            connection.Publish(msg, new CompletionListener());

            try
            {
                Thread.Sleep(1000/(rate > 0 ? rate : 1));
            } catch (Exception){}
        }
    }

    public void Close()
    {
        // Asynchronously disconnect from the eFTL server.
        connection.Disconnect();
    }

    public static void Main (string []args) 
    {
        for (int i = 0; i < args.Length; i++) 
        {
            if (args[i].Equals("--username") || args[i].Equals("-u")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    username = args[++i];
                } else {
                    Usage();
                }
            } else if (args[i].Equals("--password") || args[i].Equals("-p")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    password = args[++i];
                } else {
                    Usage();
                }
            } else if (args[i].Equals("--count") || args[i].Equals("-c")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    count = Int32.Parse(args[++i]);
                } else {
                    Usage();
                }
            } else if (args[i].Equals("--rate") || args[i].Equals("-r")) {
                if (i+1 < args.Length && !args[i+1].StartsWith("-")) {
                    rate = Int32.Parse(args[++i]);
                } else {
                    Usage();
                }
            } else if (args[i].StartsWith("-")) {
                Usage();
            } else {
                url = args[i];
            }
        }

        Console.Write("#\n# Publisher\n#\n# {0}\n#\n", EFTL.GetVersion());

        try
        {
            Publisher client = new Publisher(url, username, password);
    
            try {
                if (client.IsConnected()) {
                    client.Publish(count, rate);
                }
            } finally {
                client.Close();
            }
        } 
        catch (Exception e)
        {
            Console.WriteLine(e.ToString());
        }
    }
}
